### ISSUE TITLE
incorrect suggestion for clone_on_copy when clone is borrowed mutably

### ISSUE THREAD

**Issue by tspiteri** [rust-lang-nursery/rust-clippy#1925]
For the reduced test case below, clippy suggests changing `i.clone()` into `*i`, but that results in a "cannot borrow as mutable" error. I think `clone_on_copy` should not be triggered when the clone is borrowed mutably.

```rust
pub fn dec_read_dec(i: &mut i32) -> i32 {
    *i -= 1;
    let ret = *i;
    *i -= 1;
    ret
}

pub fn minus_1(i: &i32) -> i32 {
    dec_read_dec(&mut i.clone())
}

#[test]
fn test() {
    let mut i = 10;
    assert_eq!(minus_1(&i), 9);
    assert_eq!(i, 10);
    assert_eq!(dec_read_dec(&mut i), 9);
    assert_eq!(i, 8);
}
```

**Comment by oli-obk** [comment]
Same issue for 

```rust
    let ref mut foo = *i;
```

Both cases should be checkable by obtaining the parent of the clone expression and checking for these two special cases.
